walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,153 @@
+      for (String optionalMeetingGuest: optionalMeetingGuests) {
+        if (eventGuests.contains(optionalMeetingGuest)) {
+          unavailableTimesForOptionalGuests.add(eventTime);
+        }
+      }
+      for (String mandatoryMeetingGuest: mandatoryMeetingGuests) {
+        if (eventGuests.contains(mandatoryMeetingGuest)) {
+          fixSchedulingConflicts(potentialMeetingTimes, eventTime);
+        }
+      }
+    }
+    
+    // Make a copy of good meeting times we have at this point for mandatory guests as a backup, in case scheduling is impossible with optional guests
+    ArrayList<TimeRange> copyOfGoodMeetingTimes = new ArrayList<TimeRange>(potentialMeetingTimes);
+    // Make a hashmap of all eliminated times and frequency of elimination, to optimize scheduling with the most optional guests 
+    HashMap<TimeRange, Integer> eliminatedTimes = new HashMap<>();
+ 
+    for (TimeRange unavailableTime: unavailableTimesForOptionalGuests) {
+      // boolean to check if an unavailable time has already been deleted
+      boolean timeAccounted = false;
+      // create a list iterator for potential meeting times so it can be modified while iterating 
+      ListIterator<TimeRange> litr = potentialMeetingTimes.listIterator();
+      while (litr.hasNext()) {
+        TimeRange potentialTime = litr.next();
+        if (unavailableTime.overlaps(potentialTime)) {
+          litr.remove();
+          // put unavailable time in eliminated times hasmap and assign frequency of 1 first time it is deleted
+          eliminatedTimes.put(potentialTime, 1);
+          if (potentialTime.start() < unavailableTime.start()) {
+            TimeRange frontBlock = TimeRange.fromStartEnd(potentialTime.start(), unavailableTime.start(), false);
+            litr.add(frontBlock);
+          }
+          if (unavailableTime.end() < potentialTime.end()) {
+            TimeRange backBlock = TimeRange.fromStartEnd(unavailableTime.end(), potentialTime.end(), false);
+            litr.add(backBlock);
+          }
+        }
+        timeAccounted = true; 
+      }
+      // if the uunavailable time is not deleted, check if it has previously been deleted and inrease frequency by 1
+      if (!timeAccounted) {
+          for (TimeRange deletedTime: eliminatedTimes.keySet()) {
+              if (unavailableTime.overlaps(deletedTime)) {
+                  int count = eliminatedTimes.get(deletedTime);
+                  eliminatedTimes.replace(deletedTime, count + 1);
+              }
+          }
+      }
+    }
+    
+    // if there are no mandatory meeting guests, then we want to schedule only around optional guests
+    if (mandatoryMeetingGuests.isEmpty()) {
+      copyOfGoodMeetingTimes = potentialMeetingTimes;
+    }
+
+    // if there is potential meeting time has a shorter duration than the meeting request, then delete that potential time because it doesn't work
+    ListIterator<TimeRange> litr = potentialMeetingTimes.listIterator();
```

Comment:
Can we also rename this `litr`?

Also totally optional rabbit-hole if you're curious, but this is a place where Java Streams and filter() would make this a nice one liner:
https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html